### PR TITLE
[FIX] delivery: duplicate delivery line for real cost delivery method

### PR DESCRIPTION
**Pull request by caburj**
In de88a727b22fdeb0ab6ae332b7e9182e61f94803, it was established that
when adding delivery line that has invoice_policy == 'real', the
sale.order.line for the delivery should have price_unit == 0. However,
after 6173dd21415ab441caaf4404bf812ce2413e29f0, this feature does not
work. Setting the carrier_id in the sale.order before calling the
set_delivery_line function fixes this issue.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/delivery/models/sale_order.py [hunk 1]
```diff
@@ -48,7 +48,8 @@ def set_delivery_line(self, carrier, amount):
         self._remove_delivery_line()
 
         for order in self:
-            order._create_delivery_line(carrier, amount, price_unit_in_description=self.carrier_id.invoice_policy == 'real')
+            order.carrier_id = carrier.id
```
Comment (pimodoo) [inline addons/delivery/models/sale_order.py]: you can remove this commit, it has been merged in saas-12.3 and just wait forwardport: https://github.com/odoo/odoo/pull/35855

Comment (caburj) [reply addons/delivery/models/sale_order.py]: Done.

File: addons/delivery/wizard/choose_delivery_carrier.py [hunk 2]
```diff
@@ -67,10 +67,9 @@ def _compute_available_carrier(self):
     def _get_shipment_rate(self):
         vals = self.carrier_id.rate_shipment(self.order_id)
         if vals.get('success'):
-            if vals.get('warning_message'):
-                self.delivery_message = vals['warning_message']
+            self.delivery_message = vals.get('warning_message')
             self.delivery_price = vals['price']
-            self.display_price = vals['carrier_price']
```
Comment (pimodoo) [inline addons/delivery/wizard/choose_delivery_carrier.py]: I don't think we should change the display price, as it would mean that it is useless

Comment (pimodoo) [reply addons/delivery/wizard/choose_delivery_carrier.py]: We should porbably change it in the button_confirm function

File: addons/delivery/wizard/choose_delivery_carrier.py [hunk 3]
```diff
@@ -67,10 +67,9 @@ def _compute_available_carrier(self):
     def _get_shipment_rate(self):
         vals = self.carrier_id.rate_shipment(self.order_id)
         if vals.get('success'):
-            if vals.get('warning_message'):
-                self.delivery_message = vals['warning_message']
+            self.delivery_message = vals.get('warning_message')
```
Comment (pimodoo) [inline addons/delivery/wizard/choose_delivery_carrier.py]: vals.get('warning_message', False)

File: addons/delivery/models/sale_order.py [hunk 4]
```diff
@@ -48,7 +48,7 @@ def set_delivery_line(self, carrier, amount):
         self._remove_delivery_line()
 
         for order in self:
-            order._create_delivery_line(carrier, amount, price_unit_in_description=self.carrier_id.invoice_policy == 'real')
+            order._create_delivery_line(carrier, amount)
```
Comment (pimodoo) [inline addons/delivery/models/sale_order.py]: write commit message explaining that the carrier can be different than the one set on the order, and so we should use the correct setting of the carrier on which we are calling the request.

**Pull request closed without merging**